pkg/util/scheduler/queue_test.go

```diff
@@ -0,0 +1,635 @@
+package scheduler
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/stretchr/testify/require"
+)
+
+func QueueOptionsWithDefaults(opts *QueueOptions) *QueueOptions {
+	if opts == nil {
+		opts = &QueueOptions{}
+	}
+	if opts.MaxSizePerTenant <= 0 {
+		opts.MaxSizePerTenant = 10
+	}
+	if opts.QueueLength == nil {
+		opts.QueueLength = promauto.With(nil).NewGaugeVec(prometheus.GaugeOpts{}, []string{"slug"})
+	}
+	if opts.DiscardedRequests == nil {
+		opts.DiscardedRequests = promauto.With(nil).NewCounterVec(prometheus.CounterOpts{}, []string{"slug", "reason"})
+	}
+	if opts.EnqueueDuration == nil {
+		opts.EnqueueDuration = promauto.With(nil).NewHistogram(prometheus.HistogramOpts{})
+	}
+	return opts
+}
+
+func TestQueue(t *testing.T) { //nolint:gocyclo
+	t.Parallel()
+
+	t.Run("SimpleEnqueueAndDequeue", func(t *testing.T) {
+		t.Parallel()
+
+		ctx := context.Background()
+		q := NewQueue(QueueOptionsWithDefaults(nil))
+		require.NoError(t, q.StartAsync(ctx), "Queue should start")
+		require.NoError(t, q.AwaitRunning(ctx), "Queue should be running")
+
+		// Defer close and stopwait to ensure cleanup even if test fails
+		// Order matters: Close needs to be called before StopWait
+		defer func() {
+			q.StopAsync()
+			require.NoError(t, q.AwaitTerminated(context.Background()), "Queue should stop")
```

Comment:
You can replace these two lines with `require.NoError(t, services.StopAndAwaitTerminated(context.Background(), q))`. Same comment applies elsewhere in the test.